include/nbl/asset/ISpecializedShader.h

```diff
@@ -105,49 +78,47 @@ class ISpecializedShader : public virtual core::IReferenceCounted
-					entryPoint{_entryPoint}, shaderStage{_ss}, m_filePathHint(_filePathHint)
+				SInfo(core::smart_refctd_dynamic_array<SMapEntry>&& _entries,
+					core::smart_refctd_ptr<ICPUBuffer>&& _backingBuff,
+					const std::string& _entryPoint)
+					: entryPoint{_entryPoint}
 				{
 					setEntries(std::move(_entries),std::move(_backingBuff));
 				}
 				~SInfo() = default;
 
 				bool operator<(const SInfo& _rhs) const
 				{
-					if (shaderStage==_rhs.shaderStage)
+					if (entryPoint==_rhs.entryPoint)
 					{
-						if (entryPoint==_rhs.entryPoint)
+						size_t lhsSize = m_entries ? m_entries->size():0ull;
+						size_t rhsSize = _rhs.m_entries ? _rhs.m_entries->size():0ull;
+						if (lhsSize==rhsSize)
 						{
-							size_t lhsSize = m_entries ? m_entries->size():0ull;
-							size_t rhsSize = _rhs.m_entries ? _rhs.m_entries->size():0ull;
-							if (lhsSize==rhsSize)
+							for (size_t i=0ull; i<lhsSize; ++i)
 							{
-								for (size_t i=0ull; i<lhsSize; ++i)
-								{
-									const auto& l = (*m_entries)[i];
-									const auto& r = (*_rhs.m_entries)[i];
+								const auto& l = (*m_entries)[i];
+								const auto& r = (*_rhs.m_entries)[i];
 
-									if (l.specConstID==r.specConstID)
+								if (l.specConstID==r.specConstID)
+								{
+									if (l.size==r.size)
 									{
-										if (l.size==r.size)
-										{
-											int cmp = memcmp(reinterpret_cast<const uint8_t*>(m_backingBuffer->getPointer())+l.offset, reinterpret_cast<const uint8_t*>(_rhs.m_backingBuffer->getPointer())+r.offset, l.size);
-											if (cmp==0)
-												continue;
-											return cmp<0;
-										}
-										return l.size<r.size;
+										int cmp = memcmp(reinterpret_cast<const uint8_t*>(m_backingBuffer->getPointer())+l.offset, reinterpret_cast<const uint8_t*>(_rhs.m_backingBuffer->getPointer())+r.offset, l.size);
+										if (cmp==0)
+											continue;
+										return cmp<0;
 									}
-									return l.specConstID<r.specConstID;
+									return l.size<r.size;
 								}
-								// all entries equal if we got out the loop
-								// return m_filePathHint<_rhs.m_filePathHint; // don't do this cause OpenGL program cache might get more entries in it (I think it contains only already include-resolved shaders)
+								return l.specConstID<r.specConstID;
 							}
-							return lhsSize<rhsSize;
+							// all entries equal if we got out the loop
+							// return m_filePathHint<_rhs.m_filePathHint; // don't do this cause OpenGL program cache might get more entries in it (I think it contains only already include-resolved shaders)
 						}
-						return entryPoint<_rhs.entryPoint;
+						return lhsSize<rhsSize;
 					}
-					return shaderStage<_rhs.shaderStage;
+					return entryPoint<_rhs.entryPoint;
 				}
```

Comment:
`bool COpenGLPipelineCache::SCacheKey::operator<(const COpenGLPipelineCache::SCacheKey& _rhs) const`
https://github.com/Devsh-Graphics-Programming/Nabla/blob/vulkan_backend/src/nbl/video/COpenGLPipelineCache.cpp#L65

Reply:
@Erfan-Ahmadi please check that the OpenGL pipeline cache works, and doesn't collide on shaders that have the same shader stage

(cause @achalpandeyy removed the comparison of shader stages from the `operator<` which I think will fuck stuff up)

I think we might need to add a comparison of unspecialized shader's stages before line 69 here
https://github.com/Devsh-Graphics-Programming/Nabla/blob/vulkan_backend/src/nbl/video/COpenGLPipelineCache.cpp#L69

Reply:
@devshgraphicsprogramming 
I investigated the pipeline caching and it seems like we also check the **Spirv Code Hash**.

It doesn't get to this `operator<` for ``ISpecializedShader::SInfo`` unless we get some hash collision for different spirv codes and AFAIK the shader stage is embedded in spirv code like this: `OpEntryPoint [Fragment/Vertex/Compute] main blah blah`.

But IF we have a hash collision for some reason:
- And the shader stages are different and entrypoints have the same name with the same or no SpecializationConstants
-  And the descriptorLayouts are the same
Then it will fuck up stuff.

But If you want to check `shader stages` anyways then we must add another variable in `SCacheKey` for ShaderStage.
Let me know about you decision.

```c
	struct SCacheKey {
		std::array<uint64_t, 4> hash;
		COpenGLSpecializedShader::SInfo info;
		core::smart_refctd_ptr<COpenGLPipelineLayout> layout;

		bool operator<(const SCacheKey& _rhs) const;
	};
```

Reply:
entry points must always be "main" XD shaderc limitation.

Well then, add the stage to the `SCacheKey`